secure_message/repository/retriever.py

```diff
@@ -46,15 +52,73 @@ def retrieve_message_list(page, limit, user, ru_id=None, survey=None, cc=None, c
-                                 .label('max_date'))\
+                                 .label('max_date')) \
                 .join(Events).join(Status).outerjoin(Actors) \
                 .filter(and_(*conditions)) \
                 .filter(and_(*status_conditions)) \
                 .filter(or_(Events.event == EventsApi.SENT.value, Events.event == EventsApi.DRAFT_SAVED.value)) \
                 .group_by(SecureMessage.msg_id).subquery('t')
 
             if descend:
-                result = SecureMessage.query\
+                result = SecureMessage.query \
+                    .filter(SecureMessage.msg_id == t.c.msg_id) \
+                    .order_by(t.c.max_date.desc()).paginate(page, limit, False)
+
+            else:
+                result = SecureMessage.query \
+                    .filter(SecureMessage.msg_id == t.c.msg_id) \
+                    .order_by(t.c.max_date.asc()).paginate(page, limit, False)
+
+        except Exception as e:
+            logger.error('Error retrieving messages from database', error=e)
+            raise InternalServerError(description="Error retrieving messages from database")
+
+        return True, result
+
+    @staticmethod
+    def _retrieve_message_list_internal(page, limit, ru_id, survey, cc, ce, label, descend):
+        """returns list of messages from db"""
+        conditions = []
+        status_reject_conditions = []
+        valid_statuses = []
+        actor_conditions = []
+
+        if label is not None:
+            valid_statuses.append(label)
+            if label in [Labels.INBOX.value, Labels.ARCHIVE.value, Labels.UNREAD.value]:
+                actor_conditions.append(Actors.sent_from_internal == False)  # NOQA pylint:disable=singleton-comparison
+            if label in [Labels.DRAFT.value, Labels.SENT.value]:
+                actor_conditions.append(Actors.sent_from_internal == True)  # NOQA pylint:disable=singleton-comparison
+        else:
+            status_reject_conditions.append(Labels.DRAFT_INBOX.value)
+            valid_statuses = [Labels.INBOX.value, Labels.DRAFT.value]
+            actor_conditions.append(True)
+
+        if ru_id is not None:
+            conditions.append(SecureMessage.ru_id == str(ru_id))
+
+        if survey is not None:
+            conditions.append(SecureMessage.survey == str(survey))
+
+        if cc is not None:
+            conditions.append(SecureMessage.collection_case == str(cc))
+
+        if ce is not None:
+            conditions.append(SecureMessage.collection_exercise == str(ce))
+
+        try:
+            t = db.session.query(SecureMessage.msg_id, func.max(Events.date_time)  # pylint:disable=no-member
+                                 .label('max_date')) \
+                .join(Events).join(Status).outerjoin(Actors) \
+                .filter(and_(*conditions)) \
+                .filter(or_(*actor_conditions)) \
+                .filter(~Status.label.in_(status_reject_conditions)) \
```

Comment:
Might be worth a comment here that SQLAlchemy overrides the unary bitwise operator

Reply:
Done